### DISCUSSION TITLE
A question about the accuracy of the gradient

### DISCUSSION THREAD

**Discussion by WZDDZFJ** [mdolab/dafoam#714] [General]
In the `verification` section, you introduced forward-mode differentiation to check the correctness of the gradients. Could you provide an example of using finite differences to directly check the correctness of the gradients? Because I followed the tutorial to analyze the two-dimensional airwing of NACA0012, but when I looked at the gradient file `totalderivist.txt`, I found that the FFD points in a symmetric position had different gradient values, and the gap was relatively small when it was not deformed, but the gap was large after deformation. For example, what I've framed in red is a typical set of symmetric FFD points. Usually, we use the finite difference method to verify the gradient, that is, manually give a disturbance to the FFD point, see how the objective function changes, and then obtain the gradient value. But in this case of the airfoil, do I perturb one FFD point at a time, or do I perturb a set of symmetric FFD points at the same time and then compare it to the gradient obtained by the discrete adjoint solution?
![image](https://github.com/user-attachments/assets/3d21e737-99df-48dd-bf56-7236c75a623b)
![1733314036537](https://github.com/user-attachments/assets/5aa46005-ceaf-4689-af88-5940d2315721)

**Comment by sll426** [comment]
Hi, how to get the gradient of particular design variables value

**Reply by WZDDZFJ** [reply]
maybe in the 'totalderivist.txt'

**Comment by WZDDZFJ** [comment]
In addition to the previously asked question about the gradient of symmetric points being different, I have read your article, which contains the term `FD reference,` and I would like to ask how this came about. Referring to one of your previous replies:

> There is no automatic way to do FD for field inversion because there are so many inputs. What you can do is to manually perturb one beta value in the runScript.py file and run the primal to get the FD derivative. Check where you prescribe the initial values for beta in the runScript.py file. You can prescribe some perturbed beta values there.

My approach was to modify the code in `runScript_v2` from `pts = DVGeo.getLocalIndex(iVol) indexList = pts[:, :, :].flatten()` to `indexList = [pts[0,0,0]]` (for example) , and then use `DVGeo.addLocalDV("shapey", lower=1e-5, upper=1e-5, axis="y", scale=1.0, pointSelect=PS)`, which perturbs a specific FFD point. After calculating the flow field and subtracting the original flow field, I divide by the perturbation value to obtain the total derivative of the objective function with respect to the design variable. However, the result I obtained does **not match** the value in `totalderisvist.txt`, but the derivative I got this way is **similar** to what I get when I set `verify=3` in the `snopt optOption`. Again, my result is similar to that obtained by setting `useAD": {"mode": "fd"}`. The image below is the one I got using the same case but setting `useAD": {"mode": "fd"}`.
![image](https://github.com/user-attachments/assets/2ed0521e-29ac-4cb2-95bc-c2d915be4ace)

So to summarize the results of my experiment, this example is from the `tutorials` file. The final-difference result I calculated manually was **similar** to that of `useAD": {"mode": "fd"}` or `verify=3` in `snopt optOptions`, but **completely different** from `reverseAD`. And I also tested `forwordAD`, and the error was also **large**.
The accuracy of the gradient is very important, and different results in the same case make me very confused. **If my approach is incorrect, could you please let me know the correct way to obtain the "FD reference" similar to what is presented in your paper?** My question is a little long, thank you Professor He for your patient reply!@friedenhe

**Comment by friedenhe** [comment]
@WZDDZFJ You can use task=checkTotal (https://github.com/DAFoam/tutorials/blob/607a4f7738d067dd0921eecd4a96750c69443811/NACA0012_Airfoil/incompressible/runScript.py#L271) in a v3 script to use FD values as references. But in the case of 2D airfoil, you can't simply perturb one FFD in FD. This is because a small perturbation of any FFD point will violate the symmetry boundary condition and return unphysical results. So the FD-based totals are not correct. What you can do is use the twist (refer to a 3D wing case on how to add a twist DV into the NACA0012 tutorial) as the design variable and then run checkTotals.

**Reply by WZDDZFJ** [reply]
@friedenhe I still have a question about the gradient to ask you. On a two-dimensional airfoil, when it is not deformed, the gradient values of the FFD points symmetrical in the wingspan direction are very close. I can understand this. But why do they obtain different gradient values after synchronous deformation (that is, the result of calculating the gradient after the second time)? For example, in the above figure, the derivative value of one FFD point is 0.471, and the derivative value of the other FFD point symmetrical to it is -0.425. Then, when these two FFD points move synchronously, what should be the derivative of the objective function with respect to the design variable. Is the variation of the objective function the sum of the products of the gradients and the step sizes of the two points? I can't understand this. They have exactly the same amount of deformation. **Why are the gradient values of the symmetrical FFD points significantly different after deformation? Is this also the reason for the boundary condition?** Then how would gradient optimizers with different symmetrical FFD points be referred to？This question has puzzled me for a long time. I'm looking forward to your detailed answer

**Comment by hzrssb** [comment]
@WZDDZFJ  I also have doubts about the different gradient results of the two-dimensional airfoil symmetrical FFD points obtained in the AD mode. I find that even positive and negative are sometimes different. I guess it's a problem where two symmetrical boundaries interfere with each other, but I'm also looking forward to getting a more correct reply. @friedenhe

**Comment by friedenhe** [comment]
@WZDDZFJ and @hzrssb We also noticed this issue. When the FFD design variables are non-zero, the derivatives for the symmetry points are not the same. We suspect this is somehow related to how pyGeo handles the FFD points, but haven't gotten a chance to get to the bottom of this issue. That is why we use a different way to setup FFD points for 2D cases in the latest tutorials. Check [here](https://github.com/DAFoam/tutorials/blob/fa9c30bfb5b3af033e2a7aff9bafb17d8a248d24/NACA0012_Airfoil/incompressible/runScript.py#L138). Basically, we move the points in k=0 and k=1 simultaneously. So, there is no need to setup the symmetry constraint.

**Reply by hzrssb** [reply]
@friedenhe If the symmetrical FFD points move synchronously, is the true derivative value their sum

**Reply by hzrssb** [reply]
ADflow also uses this logic. I don't know if there is the same problem. I don't know if you think the derivative of the FFD point is accurate for the two-dimensional case?

**Reply by jxz21004** [reply]
> ADflow also uses this logic. I don't know if there is the same problem. I don't know if you think the derivative of the FFD point is accurate for the two-dimensional case?

I also noticed similar phenomena when using ADflow and IDWarp to optimize RAE2822, but the worst discrepancy I observed only reached the level at iteration 001 of the case displayed here. I suspect that this might partially originate from IDWarp. However, moving a single FFD point could make the boundary conditions somewhat unnatural.

**Reply by friedenhe** [reply]
@jxz21004 It is good to confirm that ADflow also has this issue. So, it is likely caused by IDWarp. BTW: We no longer use the shape variable in 2D airfoil cases. Instead, we use the shapeFunction, which can move 2 spanwise FFD points at the same time. So, this is no longer a problem.

**Reply by jxz21004** [reply]
> It is good to confirm that ADflow also has this issue. So, it is likely caused by IDWarp. BTW: We no longer use the shape variable in 2D airfoil cases. Instead, we use the shapeFunction, which can move 2 spanwise FFD points at the same time. So, this is no longer a problem.

Thank you very much for the clarification.

**Comment by friedenhe** [comment]
@hzrssb The deriv should be accurate. Please run check_totals using the latest NACA0012 tutorial. Also refer to the link I sent above. We haven't tested if ADflow has a similar problem.

**Reply by hzrssb** [reply]
Do you mean that the accuracy of FFD points can only be indirectly proved by testing the gradient accuracy of the torsion variable, right? Even if the derivatives of the shape variables at symmetrical FFD points are different, this might just be a logical problem of pyGeo.

**Reply by friedenhe** [reply]
> Do you mean that the accuracy of FFD points can only be indirectly proved by testing the gradient accuracy of the torsion variable, right? 

What is a torsion variable?

> Even if the derivatives of the shape variables at symmetrical FFD points are different, this might just be a logical problem of pyGeo.

Yes, that is what we suspect.

**Reply by hzrssb** [reply]
Sorry, what I mean is to use the twist of the airfoil generated by the FFD point to test the accuracy of the derivative of the FFD point.

**Comment by friedenhe** [comment]
@hzrssb No, you can still use the shape variable to verify the gradient accuracy. Check this post: https://github.com/mdolab/dafoam/discussions/818#discussioncomment-13039606 We changed the way the shape variable is set in all v4 tutorials.

**Reply by hzrssb** [reply]
I'm very sorry. Maybe the meaning I expressed was not correct enough. What I want to ask is whether the derivative values of the two-dimensional airfoil symmetrical FFD points are different in both V3 and V4 versions. Then you think this is a problem with the implementation of pyGeo, and you still think the derivative value is accurate.  I accidentally mistook this problem for not existing in the V4 version. I apologize again for my unclear expression.@friedenhe

**Comment by friedenhe** [comment]
@hzrssb Yeah, we tested this, and it seems that this problem exists in both V3 and V4.

**Reply by hzrssb** [reply]
Thank you very much for your patient reply. In fact, I think the two-dimensional nature of DAFoam is actually a kind of "pseudo-three-dimensional", but why doesn't three-dimensional have this problem? If you have any updated explanations, please do reply.

**Reply by friedenhe** [reply]
>  but why doesn't three-dimensional have this problem? 

In 3D cases, we don't need to match the displacement for two sets of FFD points..
